
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../usage_diffexp/">
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.4">
    
    
      
        <title>Test page - Repseq Library for Immune repertoire postanalysis</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.8608ea7d.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="purple">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#page-2" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Repseq Library for Immune repertoire postanalysis" class="md-header__button md-logo" aria-label="Repseq Library for Immune repertoire postanalysis" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Repseq Library for Immune repertoire postanalysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Test page
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="purple"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="lime"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../modules/" class="md-tabs__link">
        
  
    
  
  Modules

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../functions/" class="md-tabs__link">
        
  
    
  
  Functions

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../usage_mixcr/" class="md-tabs__link">
          
  
  Usage

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="./" class="md-tabs__link">
        
  
    
  
  Test page

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Repseq Library for Immune repertoire postanalysis" class="md-nav__button md-logo" aria-label="Repseq Library for Immune repertoire postanalysis" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Repseq Library for Immune repertoire postanalysis
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Modules
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../functions/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Functions
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Usage
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_mixcr/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MiXCR
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Stats
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_intersections/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Intersections
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_clustering/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clustering
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_diffexp/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Differential expression
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Test page
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Test page
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#code-annotation-examples" class="md-nav__link">
    <span class="md-ellipsis">
      Code annotation examples
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icons-and-emojis" class="md-nav__link">
    <span class="md-ellipsis">
      Icons and Emojis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#try-to-show-all-the-functions-of-a-module" class="md-nav__link">
    <span class="md-ellipsis">
      Try to show all the functions of a module
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#intersections" class="md-nav__link">
    <span class="md-ellipsis">
      intersections
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#intersections.count_table" class="md-nav__link">
    <span class="md-ellipsis">
      count_table
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#intersections.count_table_by_cluster" class="md-nav__link">
    <span class="md-ellipsis">
      count_table_by_cluster
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#intersections.find_intersecting_clonotypes" class="md-nav__link">
    <span class="md-ellipsis">
      find_intersecting_clonotypes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#intersections.intersect_clones_in_samples_batch" class="md-nav__link">
    <span class="md-ellipsis">
      intersect_clones_in_samples_batch
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#intersections.overlap_distances" class="md-nav__link">
    <span class="md-ellipsis">
      overlap_distances
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#intersections.prepare_clonotypes_dfs_for_intersections" class="md-nav__link">
    <span class="md-ellipsis">
      prepare_clonotypes_dfs_for_intersections
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="page-2">Page 2</h1>
<h2 id="code-annotation-examples">Code annotation examples</h2>
<h3 id="code-blocks">Code blocks</h3>
<p>Some <code>code</code> goes here</p>
<h3 id="plain-code-blocks">Plain code blocks</h3>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>def read_json_report(sample_id, folder, report_type):
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    filename = os.path.join(folder, f&quot;{sample_id}.{report_type}.report.json&quot;)
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    with open(filename) as data_file:    
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>        for jsonObj in data_file:
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>            report = json.loads(jsonObj)
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>    return report
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>// some comment
</code></pre></div>
<h4 id="code-for-a-specific-language">Code for a specific language</h4>
<p>Some more code with the <code>py</code> at the start</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="n">REPSEQ_PATH</span> <span class="o">=</span> <span class="s1">&#39;/home/mmyshkin/soft/repseq&#39;</span>
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a><span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">REPSEQ_PATH</span><span class="p">)</span>
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">slurm</span>
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">io</span>
<a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">common_functions</span> <span class="k">as</span> <span class="n">cf</span>
<a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">clonosets</span> <span class="k">as</span> <span class="n">cl</span>
<a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">clustering</span>
<a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">mixcr</span> <span class="k">as</span> <span class="n">mx</span>
<a id="__codelineno-1-10" name="__codelineno-1-10" href="#__codelineno-1-10"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">segment_usage</span> <span class="k">as</span> <span class="n">su</span>
<a id="__codelineno-1-11" name="__codelineno-1-11" href="#__codelineno-1-11"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">stats</span>
</code></pre></div>
<h4 id="code-with-a-title">Code with a title</h4>
<div class="highlight"><span class="filename">import useful packages</span><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a><span class="kn">from</span><span class="w"> </span><span class="nn">IPython.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">Image</span><span class="p">,</span> <span class="n">display</span>
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a><span class="kn">import</span><span class="w"> </span><span class="nn">json</span>
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a><span class="kn">import</span><span class="w"> </span><span class="nn">re</span>
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a><span class="kn">import</span><span class="w"> </span><span class="nn">math</span>
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a><span class="kn">import</span><span class="w"> </span><span class="nn">random</span>
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
</code></pre></div>
<h4 id="add-line-numbers">Add line numbers</h4>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-3-1">1</a></span>
<span class="normal"><a href="#__codelineno-3-2">2</a></span>
<span class="normal"><a href="#__codelineno-3-3">3</a></span>
<span class="normal"><a href="#__codelineno-3-4">4</a></span>
<span class="normal"><a href="#__codelineno-3-5">5</a></span>
<span class="normal"><a href="#__codelineno-3-6">6</a></span>
<span class="normal"><a href="#__codelineno-3-7">7</a></span>
<span class="normal"><a href="#__codelineno-3-8">8</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1"></a><span class="k">def</span><span class="w"> </span><span class="nf">shannon_wiener</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">):</span>
<a id="__codelineno-3-2" name="__codelineno-3-2"></a>    <span class="n">list_of_numbers</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">)</span>
<a id="__codelineno-3-3" name="__codelineno-3-3"></a>    <span class="n">total_size</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">)</span>
<a id="__codelineno-3-4" name="__codelineno-3-4"></a>    <span class="n">freqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span><span class="o">/</span><span class="n">total_size</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">list_of_numbers</span><span class="p">]</span>
<a id="__codelineno-3-5" name="__codelineno-3-5"></a>    <span class="n">diversity</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">)</span>
<a id="__codelineno-3-6" name="__codelineno-3-6"></a>    <span class="n">sw</span> <span class="o">=</span> <span class="o">-</span><span class="nb">sum</span><span class="p">([</span><span class="n">f</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">freqs</span><span class="p">])</span>
<a id="__codelineno-3-7" name="__codelineno-3-7"></a>    <span class="n">sw_norm</span> <span class="o">=</span> <span class="n">sw</span><span class="o">/</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">diversity</span><span class="p">)</span>
<a id="__codelineno-3-8" name="__codelineno-3-8"></a>    <span class="k">return</span> <span class="n">sw</span><span class="p">,</span> <span class="n">sw_norm</span><span class="p">,</span> <span class="n">diversity</span>
</code></pre></div></td></tr></table></div>
<h4 id="highlighting-lines">Highlighting lines</h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="k">def</span><span class="w"> </span><span class="nf">shannon_wiener</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">):</span>
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a>    <span class="n">list_of_numbers</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">)</span>
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>    <span class="n">total_size</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">)</span>
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>    <span class="n">freqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span><span class="o">/</span><span class="n">total_size</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">list_of_numbers</span><span class="p">]</span>
<a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a><span class="hll">    <span class="n">diversity</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">)</span>
</span><a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a><span class="hll">    <span class="n">sw</span> <span class="o">=</span> <span class="o">-</span><span class="nb">sum</span><span class="p">([</span><span class="n">f</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">freqs</span><span class="p">])</span>
</span><a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a><span class="hll">    <span class="n">sw_norm</span> <span class="o">=</span> <span class="n">sw</span><span class="o">/</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">diversity</span><span class="p">)</span>
</span><a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a>    <span class="k">return</span> <span class="n">sw</span><span class="p">,</span> <span class="n">sw_norm</span><span class="p">,</span> <span class="n">diversity</span>
</code></pre></div>
<h2 id="icons-and-emojis">Icons and Emojis</h2>
<p><img alt="ðŸ˜„" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/1f604.svg" title=":smile:" /></p>
<p><span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M464 256a208 208 0 1 0-416 0 208 208 0 1 0 416 0M0 256a256 256 0 1 1 512 0 256 256 0 1 1-512 0m130.7 57.9c-4.2-13.6 7.1-25.9 21.3-25.9h212.5c14.2 0 25.5 12.4 21.3 25.9C369 368.4 318.2 408 258.2 408s-110.8-39.6-127.5-94.1M144.4 192a32 32 0 1 1 64 0 32 32 0 1 1-64 0m165.8 21.7c-7.6 8.1-20.2 8.5-28.3.9s-8.5-20.2-.9-28.3c14.5-15.5 35.2-22.3 54.6-22.3s40.1 6.8 54.6 22.3c7.6 8.1 7.1 20.7-.9 28.3s-20.7 7.1-28.3-.9c-5.5-5.8-14.8-9.7-25.4-9.7s-19.9 3.8-25.4 9.7"/></svg></span></p>
<p><span class="twemoji twitter"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253"/></svg></span></p>
<p><span class="twemoji heart"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M14 20.408c-.492.308-.903.546-1.192.709q-.23.129-.463.252h-.002a.75.75 0 0 1-.686 0 17 17 0 0 1-.465-.252 31 31 0 0 1-4.803-3.34C3.8 15.572 1 12.331 1 8.513 1 5.052 3.829 2.5 6.736 2.5 9.03 2.5 10.881 3.726 12 5.605 13.12 3.726 14.97 2.5 17.264 2.5 20.17 2.5 23 5.052 23 8.514c0 3.818-2.801 7.06-5.389 9.262A31 31 0 0 1 14 20.408"/></svg></span></p>
<h2 id="try-to-show-all-the-functions-of-a-module">Try to show all the functions of a module</h2>


<div class="doc doc-object doc-module">



<a id="intersections"></a>
    <div class="doc doc-contents first">



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="intersections.count_table" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">count_table</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s1">&#39;aaV&#39;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">strict_presense</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Creates a table that shows how many times each unique clonotype appears across different clonosets. It processes a given dataset of clonotypes (clonosets_df) 
and generates a frequency/count table based on a specified overlap type.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>clonosets_df</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>contains three columns - <code>sample_id</code> and <code>filename</code> columns,
<code>filename</code> - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format
sample_id's should be all unique in this DF</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>overlap_type</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values are <code>aa</code>, <code>aaV</code>, <code>aaVJ</code>, <code>nt</code>, <code>ntV</code>, <code>ntVJ</code>. aa/nt define which CDR3 sequence
to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments
to decide if clonotypes are equal</p>
              </div>
            </td>
            <td>
                  <code>&#39;aaV&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>mismatches</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The permissible number of single-letter mismatches in clonotypes sequences 
for them to be treated similar, i.e. hamming distance.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>only_functional</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>use only functional clonotypes (do not contain stop codons or
frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to
1 after filtering of non-functional clonotypes</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>strict_presence</code></td>
            <td>
                  <code>(bool, default)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>): if set to <code>True</code> and the clonotype is not found in the clonoset, it will not be counted, even when the <code>mismatches</code> option is not set to 0.
If <code>False</code>, mismatched sequences are counted even if the exact match does not exist.
    by_freq (bool): default is <code>True</code> - this means that the intersect metric is frequency of clonotype, 
but not its count</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>by_freq</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>default is <code>True</code> - this means that the intersect metric is frequency of clonotype, 
but not its count</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>df</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dataframe containing clonotype counts with sample names as columns and all possible clonotypes (given the overlap_type) as rows.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-83"> 83</a></span>
<span class="normal"><a href="#__codelineno-0-84"> 84</a></span>
<span class="normal"><a href="#__codelineno-0-85"> 85</a></span>
<span class="normal"><a href="#__codelineno-0-86"> 86</a></span>
<span class="normal"><a href="#__codelineno-0-87"> 87</a></span>
<span class="normal"><a href="#__codelineno-0-88"> 88</a></span>
<span class="normal"><a href="#__codelineno-0-89"> 89</a></span>
<span class="normal"><a href="#__codelineno-0-90"> 90</a></span>
<span class="normal"><a href="#__codelineno-0-91"> 91</a></span>
<span class="normal"><a href="#__codelineno-0-92"> 92</a></span>
<span class="normal"><a href="#__codelineno-0-93"> 93</a></span>
<span class="normal"><a href="#__codelineno-0-94"> 94</a></span>
<span class="normal"><a href="#__codelineno-0-95"> 95</a></span>
<span class="normal"><a href="#__codelineno-0-96"> 96</a></span>
<span class="normal"><a href="#__codelineno-0-97"> 97</a></span>
<span class="normal"><a href="#__codelineno-0-98"> 98</a></span>
<span class="normal"><a href="#__codelineno-0-99"> 99</a></span>
<span class="normal"><a href="#__codelineno-0-100">100</a></span>
<span class="normal"><a href="#__codelineno-0-101">101</a></span>
<span class="normal"><a href="#__codelineno-0-102">102</a></span>
<span class="normal"><a href="#__codelineno-0-103">103</a></span>
<span class="normal"><a href="#__codelineno-0-104">104</a></span>
<span class="normal"><a href="#__codelineno-0-105">105</a></span>
<span class="normal"><a href="#__codelineno-0-106">106</a></span>
<span class="normal"><a href="#__codelineno-0-107">107</a></span>
<span class="normal"><a href="#__codelineno-0-108">108</a></span>
<span class="normal"><a href="#__codelineno-0-109">109</a></span>
<span class="normal"><a href="#__codelineno-0-110">110</a></span>
<span class="normal"><a href="#__codelineno-0-111">111</a></span>
<span class="normal"><a href="#__codelineno-0-112">112</a></span>
<span class="normal"><a href="#__codelineno-0-113">113</a></span>
<span class="normal"><a href="#__codelineno-0-114">114</a></span>
<span class="normal"><a href="#__codelineno-0-115">115</a></span>
<span class="normal"><a href="#__codelineno-0-116">116</a></span>
<span class="normal"><a href="#__codelineno-0-117">117</a></span>
<span class="normal"><a href="#__codelineno-0-118">118</a></span>
<span class="normal"><a href="#__codelineno-0-119">119</a></span>
<span class="normal"><a href="#__codelineno-0-120">120</a></span>
<span class="normal"><a href="#__codelineno-0-121">121</a></span>
<span class="normal"><a href="#__codelineno-0-122">122</a></span>
<span class="normal"><a href="#__codelineno-0-123">123</a></span>
<span class="normal"><a href="#__codelineno-0-124">124</a></span>
<span class="normal"><a href="#__codelineno-0-125">125</a></span>
<span class="normal"><a href="#__codelineno-0-126">126</a></span>
<span class="normal"><a href="#__codelineno-0-127">127</a></span>
<span class="normal"><a href="#__codelineno-0-128">128</a></span>
<span class="normal"><a href="#__codelineno-0-129">129</a></span>
<span class="normal"><a href="#__codelineno-0-130">130</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-83" name="__codelineno-0-83"></a><span class="k">def</span><span class="w"> </span><span class="nf">count_table</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">strict_presense</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-84" name="__codelineno-0-84"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-85" name="__codelineno-0-85"></a><span class="sd">    Creates a table that shows how many times each unique clonotype appears across different clonosets. It processes a given dataset of clonotypes (clonosets_df) </span>
<a id="__codelineno-0-86" name="__codelineno-0-86"></a><span class="sd">    and generates a frequency/count table based on a specified overlap type.</span>
<a id="__codelineno-0-87" name="__codelineno-0-87"></a>
<a id="__codelineno-0-88" name="__codelineno-0-88"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-89" name="__codelineno-0-89"></a><span class="sd">        clonosets_df (pd.DataFrame): contains three columns - `sample_id` and `filename` columns,</span>
<a id="__codelineno-0-90" name="__codelineno-0-90"></a><span class="sd">            `filename` - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format</span>
<a id="__codelineno-0-91" name="__codelineno-0-91"></a><span class="sd">            sample_id&#39;s should be all unique in this DF</span>
<a id="__codelineno-0-92" name="__codelineno-0-92"></a><span class="sd">        overlap_type (str): possible values are `aa`, `aaV`, `aaVJ`, `nt`, `ntV`, `ntVJ`. aa/nt define which CDR3 sequence</span>
<a id="__codelineno-0-93" name="__codelineno-0-93"></a><span class="sd">            to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments</span>
<a id="__codelineno-0-94" name="__codelineno-0-94"></a><span class="sd">            to decide if clonotypes are equal</span>
<a id="__codelineno-0-95" name="__codelineno-0-95"></a><span class="sd">        mismatches (int): The permissible number of single-letter mismatches in clonotypes sequences </span>
<a id="__codelineno-0-96" name="__codelineno-0-96"></a><span class="sd">            for them to be treated similar, i.e. hamming distance.</span>
<a id="__codelineno-0-97" name="__codelineno-0-97"></a><span class="sd">        only_functional (bool): use only functional clonotypes (do not contain stop codons or</span>
<a id="__codelineno-0-98" name="__codelineno-0-98"></a><span class="sd">            frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to</span>
<a id="__codelineno-0-99" name="__codelineno-0-99"></a><span class="sd">            1 after filtering of non-functional clonotypes</span>
<a id="__codelineno-0-100" name="__codelineno-0-100"></a><span class="sd">        strict_presence (bool, default:): if set to `True` and the clonotype is not found in the clonoset, it will not be counted, even when the `mismatches` option is not set to 0.</span>
<a id="__codelineno-0-101" name="__codelineno-0-101"></a><span class="sd">            If `False`, mismatched sequences are counted even if the exact match does not exist.</span>
<a id="__codelineno-0-102" name="__codelineno-0-102"></a><span class="sd">                by_freq (bool): default is `True` - this means that the intersect metric is frequency of clonotype, </span>
<a id="__codelineno-0-103" name="__codelineno-0-103"></a><span class="sd">            but not its count</span>
<a id="__codelineno-0-104" name="__codelineno-0-104"></a><span class="sd">        by_freq (bool): default is `True` - this means that the intersect metric is frequency of clonotype, </span>
<a id="__codelineno-0-105" name="__codelineno-0-105"></a><span class="sd">            but not its count</span>
<a id="__codelineno-0-106" name="__codelineno-0-106"></a>
<a id="__codelineno-0-107" name="__codelineno-0-107"></a><span class="sd">    Returns:</span>
<a id="__codelineno-0-108" name="__codelineno-0-108"></a><span class="sd">        df (pd.DataFrame): dataframe containing clonotype counts with sample names as columns and all possible clonotypes (given the overlap_type) as rows. </span>
<a id="__codelineno-0-109" name="__codelineno-0-109"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-110" name="__codelineno-0-110"></a>
<a id="__codelineno-0-111" name="__codelineno-0-111"></a>
<a id="__codelineno-0-112" name="__codelineno-0-112"></a>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Creating clonotypes count table</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="s2">&quot;-&quot;</span><span class="o">*</span><span class="mi">50</span><span class="p">)</span>
<a id="__codelineno-0-113" name="__codelineno-0-113"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overlap type: </span><span class="si">{</span><span class="n">overlap_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-114" name="__codelineno-0-114"></a>    <span class="n">aa</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span> <span class="o">=</span> <span class="n">overlap_type_to_flags</span><span class="p">(</span><span class="n">overlap_type</span><span class="p">)</span>
<a id="__codelineno-0-115" name="__codelineno-0-115"></a>    <span class="n">clonoset_dicts</span> <span class="o">=</span> <span class="n">convert_clonosets_to_compact_dicts</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">cl_filter</span><span class="p">,</span>
<a id="__codelineno-0-116" name="__codelineno-0-116"></a>                                                        <span class="n">overlap_type</span><span class="o">=</span><span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="n">by_freq</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="ow">not</span> <span class="nb">bool</span><span class="p">(</span><span class="n">mismatches</span><span class="p">))</span>
<a id="__codelineno-0-117" name="__codelineno-0-117"></a>    <span class="n">unique_clonotypes</span> <span class="o">=</span> <span class="n">find_unique_clonotypes_in_clonoset_dicts</span><span class="p">(</span><span class="n">clonoset_dicts</span><span class="p">)</span>
<a id="__codelineno-0-118" name="__codelineno-0-118"></a>
<a id="__codelineno-0-119" name="__codelineno-0-119"></a>    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-120" name="__codelineno-0-120"></a>    <span class="k">for</span> <span class="n">sample_id</span> <span class="ow">in</span> <span class="n">clonoset_dicts</span><span class="p">:</span>
<a id="__codelineno-0-121" name="__codelineno-0-121"></a>        <span class="n">task</span> <span class="o">=</span> <span class="p">[</span><span class="n">unique_clonotypes</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">clonoset_dicts</span><span class="p">[</span><span class="n">sample_id</span><span class="p">],</span> <span class="n">mismatches</span><span class="p">,</span> <span class="n">strict_presense</span><span class="p">]</span>
<a id="__codelineno-0-122" name="__codelineno-0-122"></a>        <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">task</span><span class="p">)</span>
<a id="__codelineno-0-123" name="__codelineno-0-123"></a>
<a id="__codelineno-0-124" name="__codelineno-0-124"></a>    <span class="n">results</span> <span class="o">=</span> <span class="n">run_parallel_calculation</span><span class="p">(</span><span class="n">count_table_mp</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="s2">&quot;Counting features&quot;</span><span class="p">,</span> <span class="n">object_name</span><span class="o">=</span><span class="s2">&quot;clonosets&quot;</span><span class="p">)</span>
<a id="__codelineno-0-125" name="__codelineno-0-125"></a>    <span class="n">result_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-126" name="__codelineno-0-126"></a>    <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
<a id="__codelineno-0-127" name="__codelineno-0-127"></a>        <span class="n">result_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<a id="__codelineno-0-128" name="__codelineno-0-128"></a>    <span class="n">count_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">result_dict</span><span class="p">)</span>
<a id="__codelineno-0-129" name="__codelineno-0-129"></a>    <span class="n">count_table</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">unique_clonotypes</span>
<a id="__codelineno-0-130" name="__codelineno-0-130"></a>    <span class="k">return</span> <span class="n">count_table</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="intersections.count_table_by_cluster" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">count_table_by_cluster</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clusters_list</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s1">&#39;aaV&#39;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>This function creates a table that shows the presence of clonotypes grouped into user-provided clusters across different clonosets. Instead of counting individual clonotypes, it 
calculates how many clonotypes from each cluster appear in each clonoset.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>clonosets_df</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>contains three columns - <code>sample_id</code> and <code>filename</code> columns,
filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format
sample_id's should be all unique in this DF</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>cluster_list</code></td>
            <td>
                  <code>?</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>description</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>overlap_type</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values are <code>aa</code>, <code>aaV</code>, <code>aaVJ</code>, <code>nt</code>, <code>ntV</code>, <code>ntVJ</code>. aa/nt define which CDR3 sequence
to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments
to decide if clonotypes are equal</p>
              </div>
            </td>
            <td>
                  <code>&#39;aaV&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>mismatches</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The permissible number of single-letter mismatches in clonotypes sequences 
for them to be treated similar, i.e. hamming distance.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>only_functional</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>use only functional clonotypes (do not contain stop codons or
frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to
1 after filtering of non-functional clonotypes</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>df</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dataframe with following columns: <code>clone</code>, <code>sample1_count</code>, <code>sample2_count</code>, <code>sample1</code>, <code>sample2</code>, <code>pair</code>. 
<code>clone</code> - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric
count columns contain freq/count of the clone in sample
pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-162">162</a></span>
<span class="normal"><a href="#__codelineno-0-163">163</a></span>
<span class="normal"><a href="#__codelineno-0-164">164</a></span>
<span class="normal"><a href="#__codelineno-0-165">165</a></span>
<span class="normal"><a href="#__codelineno-0-166">166</a></span>
<span class="normal"><a href="#__codelineno-0-167">167</a></span>
<span class="normal"><a href="#__codelineno-0-168">168</a></span>
<span class="normal"><a href="#__codelineno-0-169">169</a></span>
<span class="normal"><a href="#__codelineno-0-170">170</a></span>
<span class="normal"><a href="#__codelineno-0-171">171</a></span>
<span class="normal"><a href="#__codelineno-0-172">172</a></span>
<span class="normal"><a href="#__codelineno-0-173">173</a></span>
<span class="normal"><a href="#__codelineno-0-174">174</a></span>
<span class="normal"><a href="#__codelineno-0-175">175</a></span>
<span class="normal"><a href="#__codelineno-0-176">176</a></span>
<span class="normal"><a href="#__codelineno-0-177">177</a></span>
<span class="normal"><a href="#__codelineno-0-178">178</a></span>
<span class="normal"><a href="#__codelineno-0-179">179</a></span>
<span class="normal"><a href="#__codelineno-0-180">180</a></span>
<span class="normal"><a href="#__codelineno-0-181">181</a></span>
<span class="normal"><a href="#__codelineno-0-182">182</a></span>
<span class="normal"><a href="#__codelineno-0-183">183</a></span>
<span class="normal"><a href="#__codelineno-0-184">184</a></span>
<span class="normal"><a href="#__codelineno-0-185">185</a></span>
<span class="normal"><a href="#__codelineno-0-186">186</a></span>
<span class="normal"><a href="#__codelineno-0-187">187</a></span>
<span class="normal"><a href="#__codelineno-0-188">188</a></span>
<span class="normal"><a href="#__codelineno-0-189">189</a></span>
<span class="normal"><a href="#__codelineno-0-190">190</a></span>
<span class="normal"><a href="#__codelineno-0-191">191</a></span>
<span class="normal"><a href="#__codelineno-0-192">192</a></span>
<span class="normal"><a href="#__codelineno-0-193">193</a></span>
<span class="normal"><a href="#__codelineno-0-194">194</a></span>
<span class="normal"><a href="#__codelineno-0-195">195</a></span>
<span class="normal"><a href="#__codelineno-0-196">196</a></span>
<span class="normal"><a href="#__codelineno-0-197">197</a></span>
<span class="normal"><a href="#__codelineno-0-198">198</a></span>
<span class="normal"><a href="#__codelineno-0-199">199</a></span>
<span class="normal"><a href="#__codelineno-0-200">200</a></span>
<span class="normal"><a href="#__codelineno-0-201">201</a></span>
<span class="normal"><a href="#__codelineno-0-202">202</a></span>
<span class="normal"><a href="#__codelineno-0-203">203</a></span>
<span class="normal"><a href="#__codelineno-0-204">204</a></span>
<span class="normal"><a href="#__codelineno-0-205">205</a></span>
<span class="normal"><a href="#__codelineno-0-206">206</a></span>
<span class="normal"><a href="#__codelineno-0-207">207</a></span>
<span class="normal"><a href="#__codelineno-0-208">208</a></span>
<span class="normal"><a href="#__codelineno-0-209">209</a></span>
<span class="normal"><a href="#__codelineno-0-210">210</a></span>
<span class="normal"><a href="#__codelineno-0-211">211</a></span>
<span class="normal"><a href="#__codelineno-0-212">212</a></span>
<span class="normal"><a href="#__codelineno-0-213">213</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-162" name="__codelineno-0-162"></a><span class="k">def</span><span class="w"> </span><span class="nf">count_table_by_cluster</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clusters_list</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-163" name="__codelineno-0-163"></a>
<a id="__codelineno-0-164" name="__codelineno-0-164"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-165" name="__codelineno-0-165"></a><span class="sd">    This function creates a table that shows the presence of clonotypes grouped into user-provided clusters across different clonosets. Instead of counting individual clonotypes, it </span>
<a id="__codelineno-0-166" name="__codelineno-0-166"></a><span class="sd">    calculates how many clonotypes from each cluster appear in each clonoset.</span>
<a id="__codelineno-0-167" name="__codelineno-0-167"></a>
<a id="__codelineno-0-168" name="__codelineno-0-168"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-169" name="__codelineno-0-169"></a><span class="sd">        clonosets_df (pd.DataFrame): contains three columns - `sample_id` and `filename` columns,</span>
<a id="__codelineno-0-170" name="__codelineno-0-170"></a><span class="sd">            filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format</span>
<a id="__codelineno-0-171" name="__codelineno-0-171"></a><span class="sd">            sample_id&#39;s should be all unique in this DF</span>
<a id="__codelineno-0-172" name="__codelineno-0-172"></a><span class="sd">        cluster_list (?): description</span>
<a id="__codelineno-0-173" name="__codelineno-0-173"></a><span class="sd">        overlap_type (str): possible values are `aa`, `aaV`, `aaVJ`, `nt`, `ntV`, `ntVJ`. aa/nt define which CDR3 sequence</span>
<a id="__codelineno-0-174" name="__codelineno-0-174"></a><span class="sd">            to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments</span>
<a id="__codelineno-0-175" name="__codelineno-0-175"></a><span class="sd">            to decide if clonotypes are equal</span>
<a id="__codelineno-0-176" name="__codelineno-0-176"></a><span class="sd">        mismatches (int): The permissible number of single-letter mismatches in clonotypes sequences </span>
<a id="__codelineno-0-177" name="__codelineno-0-177"></a><span class="sd">            for them to be treated similar, i.e. hamming distance.</span>
<a id="__codelineno-0-178" name="__codelineno-0-178"></a><span class="sd">        only_functional (bool): use only functional clonotypes (do not contain stop codons or</span>
<a id="__codelineno-0-179" name="__codelineno-0-179"></a><span class="sd">            frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to</span>
<a id="__codelineno-0-180" name="__codelineno-0-180"></a><span class="sd">            1 after filtering of non-functional clonotypes</span>
<a id="__codelineno-0-181" name="__codelineno-0-181"></a>
<a id="__codelineno-0-182" name="__codelineno-0-182"></a><span class="sd">    Returns:</span>
<a id="__codelineno-0-183" name="__codelineno-0-183"></a><span class="sd">        df (pd.DataFrame): dataframe with following columns: `clone`, `sample1_count`, `sample2_count`, `sample1`, `sample2`, `pair`. </span>
<a id="__codelineno-0-184" name="__codelineno-0-184"></a><span class="sd">            `clone` - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric</span>
<a id="__codelineno-0-185" name="__codelineno-0-185"></a><span class="sd">            count columns contain freq/count of the clone in sample</span>
<a id="__codelineno-0-186" name="__codelineno-0-186"></a><span class="sd">            pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</span>
<a id="__codelineno-0-187" name="__codelineno-0-187"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-188" name="__codelineno-0-188"></a>
<a id="__codelineno-0-189" name="__codelineno-0-189"></a>
<a id="__codelineno-0-190" name="__codelineno-0-190"></a>
<a id="__codelineno-0-191" name="__codelineno-0-191"></a>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Creating clonotypes count table</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="s2">&quot;-&quot;</span><span class="o">*</span><span class="mi">50</span><span class="p">)</span>
<a id="__codelineno-0-192" name="__codelineno-0-192"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overlap type: </span><span class="si">{</span><span class="n">overlap_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-193" name="__codelineno-0-193"></a>
<a id="__codelineno-0-194" name="__codelineno-0-194"></a>    <span class="n">aa</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span> <span class="o">=</span> <span class="n">overlap_type_to_flags</span><span class="p">(</span><span class="n">overlap_type</span><span class="p">)</span>
<a id="__codelineno-0-195" name="__codelineno-0-195"></a>
<a id="__codelineno-0-196" name="__codelineno-0-196"></a>    <span class="n">clonoset_dicts</span> <span class="o">=</span> <span class="n">convert_clonosets_to_compact_dicts</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">cl_filter</span><span class="p">,</span>
<a id="__codelineno-0-197" name="__codelineno-0-197"></a>                                                        <span class="n">overlap_type</span><span class="o">=</span><span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="n">by_freq</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="ow">not</span> <span class="nb">bool</span><span class="p">(</span><span class="n">mismatches</span><span class="p">))</span>
<a id="__codelineno-0-198" name="__codelineno-0-198"></a>
<a id="__codelineno-0-199" name="__codelineno-0-199"></a>    <span class="n">clonotypes_by_cluster</span> <span class="o">=</span> <span class="n">convert_clusters_to_clonotype_list</span><span class="p">(</span><span class="n">clusters_list</span><span class="p">,</span> <span class="n">aa</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">)</span>
<a id="__codelineno-0-200" name="__codelineno-0-200"></a>
<a id="__codelineno-0-201" name="__codelineno-0-201"></a>    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-202" name="__codelineno-0-202"></a>    <span class="k">for</span> <span class="n">sample_id</span> <span class="ow">in</span> <span class="n">clonoset_dicts</span><span class="p">:</span>
<a id="__codelineno-0-203" name="__codelineno-0-203"></a>        <span class="n">task</span> <span class="o">=</span> <span class="p">[</span><span class="n">clonotypes_by_cluster</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">clonoset_dicts</span><span class="p">[</span><span class="n">sample_id</span><span class="p">],</span> <span class="n">mismatches</span><span class="p">]</span>
<a id="__codelineno-0-204" name="__codelineno-0-204"></a>        <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">task</span><span class="p">)</span>
<a id="__codelineno-0-205" name="__codelineno-0-205"></a>
<a id="__codelineno-0-206" name="__codelineno-0-206"></a>    <span class="n">results</span> <span class="o">=</span> <span class="n">run_parallel_calculation</span><span class="p">(</span><span class="n">count_table_by_cluster_mp</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="s2">&quot;Counting cluster presence&quot;</span><span class="p">,</span> <span class="n">object_name</span><span class="o">=</span><span class="s2">&quot;clonosets&quot;</span><span class="p">)</span>
<a id="__codelineno-0-207" name="__codelineno-0-207"></a>    <span class="n">result_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-208" name="__codelineno-0-208"></a>    <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
<a id="__codelineno-0-209" name="__codelineno-0-209"></a>        <span class="n">result_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<a id="__codelineno-0-210" name="__codelineno-0-210"></a>    <span class="n">count_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">result_dict</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;index&quot;</span><span class="p">:</span><span class="s2">&quot;feature_id&quot;</span><span class="p">})</span>
<a id="__codelineno-0-211" name="__codelineno-0-211"></a>    <span class="n">count_table</span><span class="p">[</span><span class="s2">&quot;feature_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">count_table</span><span class="p">[</span><span class="s2">&quot;feature_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;cluster_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-212" name="__codelineno-0-212"></a>
<a id="__codelineno-0-213" name="__codelineno-0-213"></a>    <span class="k">return</span> <span class="n">count_table</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="intersections.find_intersecting_clonotypes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">find_intersecting_clonotypes</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s1">&#39;aaV&#39;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;F2&#39;</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Calculating overlap distances between multiple repseq samples using F2 of F metrics
The result of this function may be used for heatmap+clusterization of samples or for MDS plots</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>clonosets_df</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>contains three columns - <code>sample_id</code> and <code>filename</code> columns,
filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format
sample_id's should be all unique in this DF</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>overlap_type</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values are <code>aa</code>, <code>aaV</code>, <code>aaVJ</code>, <code>nt</code>, <code>ntV</code>, <code>ntVJ</code>. aa/nt define which CDR3 sequence
to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments
to decide if clonotypes are equal</p>
              </div>
            </td>
            <td>
                  <code>&#39;aaV&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>mismatches</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The permissible number of single-letter mismatches in clonotypes sequences 
for them to be treated similar, i.e. hamming distance.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>by_umi</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>set =True for MiXCR4 clonosets to select count/frequency of clonotypes 
in UMI's if they exist in implemented protocol</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>metric</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values - <code>F</code>, <code>F2</code> or <code>C</code>. Default <code>F2</code>. <code>F2</code> - sum of sqrt of product of 
similar clonotype frequencies in two clonosets. <code>F</code> - sqrt of the sum of frequency products.
<code>C</code> - total frequency of clonotypes in <code>sample1</code>, that are similar to clonotypes in <code>sample2</code></p>
              </div>
            </td>
            <td>
                  <code>&#39;F2&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>only_functional</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>use only functional clonotypes (do not contain stop codons or
frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to
1 after filtering of non-functional clonotypes</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Important: similar clonotypes by <code>overlap_type</code> in one particular clonoset are NOT combined into one
and are treated as different clonotypes.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>df</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dataframe with following columns: <code>clone</code>, <code>sample1_count</code>, <code>sample2_count</code>, <code>sample1</code>, <code>sample2</code>, <code>pair</code>. 
<code>clone</code> - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric
count columns contain freq/count of the clone in sample
pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-479">479</a></span>
<span class="normal"><a href="#__codelineno-0-480">480</a></span>
<span class="normal"><a href="#__codelineno-0-481">481</a></span>
<span class="normal"><a href="#__codelineno-0-482">482</a></span>
<span class="normal"><a href="#__codelineno-0-483">483</a></span>
<span class="normal"><a href="#__codelineno-0-484">484</a></span>
<span class="normal"><a href="#__codelineno-0-485">485</a></span>
<span class="normal"><a href="#__codelineno-0-486">486</a></span>
<span class="normal"><a href="#__codelineno-0-487">487</a></span>
<span class="normal"><a href="#__codelineno-0-488">488</a></span>
<span class="normal"><a href="#__codelineno-0-489">489</a></span>
<span class="normal"><a href="#__codelineno-0-490">490</a></span>
<span class="normal"><a href="#__codelineno-0-491">491</a></span>
<span class="normal"><a href="#__codelineno-0-492">492</a></span>
<span class="normal"><a href="#__codelineno-0-493">493</a></span>
<span class="normal"><a href="#__codelineno-0-494">494</a></span>
<span class="normal"><a href="#__codelineno-0-495">495</a></span>
<span class="normal"><a href="#__codelineno-0-496">496</a></span>
<span class="normal"><a href="#__codelineno-0-497">497</a></span>
<span class="normal"><a href="#__codelineno-0-498">498</a></span>
<span class="normal"><a href="#__codelineno-0-499">499</a></span>
<span class="normal"><a href="#__codelineno-0-500">500</a></span>
<span class="normal"><a href="#__codelineno-0-501">501</a></span>
<span class="normal"><a href="#__codelineno-0-502">502</a></span>
<span class="normal"><a href="#__codelineno-0-503">503</a></span>
<span class="normal"><a href="#__codelineno-0-504">504</a></span>
<span class="normal"><a href="#__codelineno-0-505">505</a></span>
<span class="normal"><a href="#__codelineno-0-506">506</a></span>
<span class="normal"><a href="#__codelineno-0-507">507</a></span>
<span class="normal"><a href="#__codelineno-0-508">508</a></span>
<span class="normal"><a href="#__codelineno-0-509">509</a></span>
<span class="normal"><a href="#__codelineno-0-510">510</a></span>
<span class="normal"><a href="#__codelineno-0-511">511</a></span>
<span class="normal"><a href="#__codelineno-0-512">512</a></span>
<span class="normal"><a href="#__codelineno-0-513">513</a></span>
<span class="normal"><a href="#__codelineno-0-514">514</a></span>
<span class="normal"><a href="#__codelineno-0-515">515</a></span>
<span class="normal"><a href="#__codelineno-0-516">516</a></span>
<span class="normal"><a href="#__codelineno-0-517">517</a></span>
<span class="normal"><a href="#__codelineno-0-518">518</a></span>
<span class="normal"><a href="#__codelineno-0-519">519</a></span>
<span class="normal"><a href="#__codelineno-0-520">520</a></span>
<span class="normal"><a href="#__codelineno-0-521">521</a></span>
<span class="normal"><a href="#__codelineno-0-522">522</a></span>
<span class="normal"><a href="#__codelineno-0-523">523</a></span>
<span class="normal"><a href="#__codelineno-0-524">524</a></span>
<span class="normal"><a href="#__codelineno-0-525">525</a></span>
<span class="normal"><a href="#__codelineno-0-526">526</a></span>
<span class="normal"><a href="#__codelineno-0-527">527</a></span>
<span class="normal"><a href="#__codelineno-0-528">528</a></span>
<span class="normal"><a href="#__codelineno-0-529">529</a></span>
<span class="normal"><a href="#__codelineno-0-530">530</a></span>
<span class="normal"><a href="#__codelineno-0-531">531</a></span>
<span class="normal"><a href="#__codelineno-0-532">532</a></span>
<span class="normal"><a href="#__codelineno-0-533">533</a></span>
<span class="normal"><a href="#__codelineno-0-534">534</a></span>
<span class="normal"><a href="#__codelineno-0-535">535</a></span>
<span class="normal"><a href="#__codelineno-0-536">536</a></span>
<span class="normal"><a href="#__codelineno-0-537">537</a></span>
<span class="normal"><a href="#__codelineno-0-538">538</a></span>
<span class="normal"><a href="#__codelineno-0-539">539</a></span>
<span class="normal"><a href="#__codelineno-0-540">540</a></span>
<span class="normal"><a href="#__codelineno-0-541">541</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-479" name="__codelineno-0-479"></a><span class="k">def</span><span class="w"> </span><span class="nf">find_intersecting_clonotypes</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;F2&quot;</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-480" name="__codelineno-0-480"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-481" name="__codelineno-0-481"></a><span class="sd">    Calculating overlap distances between multiple repseq samples using F2 of F metrics</span>
<a id="__codelineno-0-482" name="__codelineno-0-482"></a><span class="sd">    The result of this function may be used for heatmap+clusterization of samples or for MDS plots</span>
<a id="__codelineno-0-483" name="__codelineno-0-483"></a>
<a id="__codelineno-0-484" name="__codelineno-0-484"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-485" name="__codelineno-0-485"></a><span class="sd">        clonosets_df (pd.DataFrame): contains three columns - `sample_id` and `filename` columns,</span>
<a id="__codelineno-0-486" name="__codelineno-0-486"></a><span class="sd">            filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format</span>
<a id="__codelineno-0-487" name="__codelineno-0-487"></a><span class="sd">            sample_id&#39;s should be all unique in this DF</span>
<a id="__codelineno-0-488" name="__codelineno-0-488"></a><span class="sd">        overlap_type (str): possible values are `aa`, `aaV`, `aaVJ`, `nt`, `ntV`, `ntVJ`. aa/nt define which CDR3 sequence</span>
<a id="__codelineno-0-489" name="__codelineno-0-489"></a><span class="sd">            to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments</span>
<a id="__codelineno-0-490" name="__codelineno-0-490"></a><span class="sd">            to decide if clonotypes are equal</span>
<a id="__codelineno-0-491" name="__codelineno-0-491"></a><span class="sd">        mismatches (int): The permissible number of single-letter mismatches in clonotypes sequences </span>
<a id="__codelineno-0-492" name="__codelineno-0-492"></a><span class="sd">            for them to be treated similar, i.e. hamming distance.</span>
<a id="__codelineno-0-493" name="__codelineno-0-493"></a><span class="sd">        by_umi (bool): set =True for MiXCR4 clonosets to select count/frequency of clonotypes </span>
<a id="__codelineno-0-494" name="__codelineno-0-494"></a><span class="sd">            in UMI&#39;s if they exist in implemented protocol</span>
<a id="__codelineno-0-495" name="__codelineno-0-495"></a><span class="sd">        metric (str): possible values - `F`, `F2` or `C`. Default `F2`. `F2` - sum of sqrt of product of </span>
<a id="__codelineno-0-496" name="__codelineno-0-496"></a><span class="sd">            similar clonotype frequencies in two clonosets. `F` - sqrt of the sum of frequency products.</span>
<a id="__codelineno-0-497" name="__codelineno-0-497"></a><span class="sd">            `C` - total frequency of clonotypes in `sample1`, that are similar to clonotypes in `sample2`</span>
<a id="__codelineno-0-498" name="__codelineno-0-498"></a><span class="sd">        only_functional (bool): use only functional clonotypes (do not contain stop codons or</span>
<a id="__codelineno-0-499" name="__codelineno-0-499"></a><span class="sd">            frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to</span>
<a id="__codelineno-0-500" name="__codelineno-0-500"></a><span class="sd">            1 after filtering of non-functional clonotypes</span>
<a id="__codelineno-0-501" name="__codelineno-0-501"></a>
<a id="__codelineno-0-502" name="__codelineno-0-502"></a><span class="sd">    Important: similar clonotypes by `overlap_type` in one particular clonoset are NOT combined into one</span>
<a id="__codelineno-0-503" name="__codelineno-0-503"></a><span class="sd">    and are treated as different clonotypes.</span>
<a id="__codelineno-0-504" name="__codelineno-0-504"></a>
<a id="__codelineno-0-505" name="__codelineno-0-505"></a><span class="sd">    Returns:</span>
<a id="__codelineno-0-506" name="__codelineno-0-506"></a><span class="sd">        df (pd.DataFrame): dataframe with following columns: `clone`, `sample1_count`, `sample2_count`, `sample1`, `sample2`, `pair`. </span>
<a id="__codelineno-0-507" name="__codelineno-0-507"></a><span class="sd">            `clone` - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric</span>
<a id="__codelineno-0-508" name="__codelineno-0-508"></a><span class="sd">            count columns contain freq/count of the clone in sample</span>
<a id="__codelineno-0-509" name="__codelineno-0-509"></a><span class="sd">            pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</span>
<a id="__codelineno-0-510" name="__codelineno-0-510"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-511" name="__codelineno-0-511"></a>
<a id="__codelineno-0-512" name="__codelineno-0-512"></a>
<a id="__codelineno-0-513" name="__codelineno-0-513"></a>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Intersecting clones in clonosets</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="s2">&quot;-&quot;</span><span class="o">*</span><span class="mi">50</span><span class="p">)</span>
<a id="__codelineno-0-514" name="__codelineno-0-514"></a>    <span class="n">aa</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span> <span class="o">=</span> <span class="n">overlap_type_to_flags</span><span class="p">(</span><span class="n">overlap_type</span><span class="p">)</span>
<a id="__codelineno-0-515" name="__codelineno-0-515"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overlap type: </span><span class="si">{</span><span class="n">overlap_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-516" name="__codelineno-0-516"></a>
<a id="__codelineno-0-517" name="__codelineno-0-517"></a>    <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">samples_total</span><span class="p">,</span> <span class="n">two_dataframes</span><span class="p">,</span> <span class="n">sample_list</span><span class="p">,</span> <span class="n">sample_list2</span> <span class="o">=</span> <span class="n">prepare_clonotypes_dfs_for_intersections</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="p">,</span>
<a id="__codelineno-0-518" name="__codelineno-0-518"></a>                                                                                                                        <span class="n">cl_filter</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="p">,</span>
<a id="__codelineno-0-519" name="__codelineno-0-519"></a>                                                                                                                        <span class="n">overlap_type</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="ow">not</span> <span class="nb">bool</span><span class="p">(</span><span class="n">mismatches</span><span class="p">))</span>
<a id="__codelineno-0-520" name="__codelineno-0-520"></a>
<a id="__codelineno-0-521" name="__codelineno-0-521"></a>    <span class="c1"># generating a set of tasks</span>
<a id="__codelineno-0-522" name="__codelineno-0-522"></a>
<a id="__codelineno-0-523" name="__codelineno-0-523"></a>    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-524" name="__codelineno-0-524"></a>
<a id="__codelineno-0-525" name="__codelineno-0-525"></a>    <span class="k">if</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-526" name="__codelineno-0-526"></a>        <span class="k">for</span> <span class="n">sample1</span> <span class="ow">in</span> <span class="n">sample_list</span><span class="p">:</span>
<a id="__codelineno-0-527" name="__codelineno-0-527"></a>            <span class="k">for</span> <span class="n">sample2</span> <span class="ow">in</span> <span class="n">sample_list2</span><span class="p">:</span>
<a id="__codelineno-0-528" name="__codelineno-0-528"></a>                <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">))</span>        
<a id="__codelineno-0-529" name="__codelineno-0-529"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-530" name="__codelineno-0-530"></a>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="p">):</span>
<a id="__codelineno-0-531" name="__codelineno-0-531"></a>            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="p">):</span>
<a id="__codelineno-0-532" name="__codelineno-0-532"></a>                <span class="n">sample1</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-533" name="__codelineno-0-533"></a>                <span class="n">sample2</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
<a id="__codelineno-0-534" name="__codelineno-0-534"></a>                <span class="k">if</span> <span class="n">sample1</span> <span class="o">!=</span> <span class="n">sample2</span><span class="p">:</span>
<a id="__codelineno-0-535" name="__codelineno-0-535"></a>                    <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">))</span>
<a id="__codelineno-0-536" name="__codelineno-0-536"></a>
<a id="__codelineno-0-537" name="__codelineno-0-537"></a>
<a id="__codelineno-0-538" name="__codelineno-0-538"></a>    <span class="c1"># run calculation in parallel</span>
<a id="__codelineno-0-539" name="__codelineno-0-539"></a>    <span class="n">result_list</span> <span class="o">=</span> <span class="n">run_parallel_calculation</span><span class="p">(</span><span class="n">find_overlapping_clones_in_two_clone_dicts</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="s2">&quot;Intersecting clonosets&quot;</span><span class="p">,</span> <span class="n">object_name</span><span class="o">=</span><span class="s2">&quot;pairs&quot;</span><span class="p">)</span>
<a id="__codelineno-0-540" name="__codelineno-0-540"></a>
<a id="__codelineno-0-541" name="__codelineno-0-541"></a>    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">result_list</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="intersections.intersect_clones_in_samples_batch" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">intersect_clones_in_samples_batch</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s1">&#39;aaV&#39;</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Calculating frequencies of intersecting clonotypes between multiple repseq samples.
The result of this function may be used for scatterplots of frequencies/counts of 
overlapping clonotypes</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>clonosets_df</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>contains three columns - <code>sample_id</code> and <code>filename</code> columns,
<code>filename</code> - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format
sample_id's should be all unique in this DF</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>overlap_type</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values are <code>aa</code>, <code>aaV</code>, <code>aaVJ</code>, <code>nt</code>, <code>ntV</code>, <code>ntVJ</code>. aa/nt define which CDR3 sequence
to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments
to decide if clonotypes are equal</p>
              </div>
            </td>
            <td>
                  <code>&#39;aaV&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>by_umi</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>set <code>=True</code> for MiXCR4 clonosets to select count/frequency of clonotypes 
in UMI's if they exist in implemented protocol</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>by_freq</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>default is <code>True</code> - this means that the intersect metric is frequency of clonotype, 
but not its count</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>only_functional</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>use only functional clonotypes (do not contain stop codons or
frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to
1 after filtering of non-functional clonotypes</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Important: when using particular overlap type, similar clonotypes in one particular clonoset are
combined into one with summation of counts/frequencies.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>df</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dataframe with following columns: <code>clone</code>, <code>sample1_count</code>, <code>sample2_count</code>, <code>sample1</code>, <code>sample2</code>, <code>pair</code>
clone - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric
count columns contain freq/count of the clone in sample
pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-20">20</a></span>
<span class="normal"><a href="#__codelineno-0-21">21</a></span>
<span class="normal"><a href="#__codelineno-0-22">22</a></span>
<span class="normal"><a href="#__codelineno-0-23">23</a></span>
<span class="normal"><a href="#__codelineno-0-24">24</a></span>
<span class="normal"><a href="#__codelineno-0-25">25</a></span>
<span class="normal"><a href="#__codelineno-0-26">26</a></span>
<span class="normal"><a href="#__codelineno-0-27">27</a></span>
<span class="normal"><a href="#__codelineno-0-28">28</a></span>
<span class="normal"><a href="#__codelineno-0-29">29</a></span>
<span class="normal"><a href="#__codelineno-0-30">30</a></span>
<span class="normal"><a href="#__codelineno-0-31">31</a></span>
<span class="normal"><a href="#__codelineno-0-32">32</a></span>
<span class="normal"><a href="#__codelineno-0-33">33</a></span>
<span class="normal"><a href="#__codelineno-0-34">34</a></span>
<span class="normal"><a href="#__codelineno-0-35">35</a></span>
<span class="normal"><a href="#__codelineno-0-36">36</a></span>
<span class="normal"><a href="#__codelineno-0-37">37</a></span>
<span class="normal"><a href="#__codelineno-0-38">38</a></span>
<span class="normal"><a href="#__codelineno-0-39">39</a></span>
<span class="normal"><a href="#__codelineno-0-40">40</a></span>
<span class="normal"><a href="#__codelineno-0-41">41</a></span>
<span class="normal"><a href="#__codelineno-0-42">42</a></span>
<span class="normal"><a href="#__codelineno-0-43">43</a></span>
<span class="normal"><a href="#__codelineno-0-44">44</a></span>
<span class="normal"><a href="#__codelineno-0-45">45</a></span>
<span class="normal"><a href="#__codelineno-0-46">46</a></span>
<span class="normal"><a href="#__codelineno-0-47">47</a></span>
<span class="normal"><a href="#__codelineno-0-48">48</a></span>
<span class="normal"><a href="#__codelineno-0-49">49</a></span>
<span class="normal"><a href="#__codelineno-0-50">50</a></span>
<span class="normal"><a href="#__codelineno-0-51">51</a></span>
<span class="normal"><a href="#__codelineno-0-52">52</a></span>
<span class="normal"><a href="#__codelineno-0-53">53</a></span>
<span class="normal"><a href="#__codelineno-0-54">54</a></span>
<span class="normal"><a href="#__codelineno-0-55">55</a></span>
<span class="normal"><a href="#__codelineno-0-56">56</a></span>
<span class="normal"><a href="#__codelineno-0-57">57</a></span>
<span class="normal"><a href="#__codelineno-0-58">58</a></span>
<span class="normal"><a href="#__codelineno-0-59">59</a></span>
<span class="normal"><a href="#__codelineno-0-60">60</a></span>
<span class="normal"><a href="#__codelineno-0-61">61</a></span>
<span class="normal"><a href="#__codelineno-0-62">62</a></span>
<span class="normal"><a href="#__codelineno-0-63">63</a></span>
<span class="normal"><a href="#__codelineno-0-64">64</a></span>
<span class="normal"><a href="#__codelineno-0-65">65</a></span>
<span class="normal"><a href="#__codelineno-0-66">66</a></span>
<span class="normal"><a href="#__codelineno-0-67">67</a></span>
<span class="normal"><a href="#__codelineno-0-68">68</a></span>
<span class="normal"><a href="#__codelineno-0-69">69</a></span>
<span class="normal"><a href="#__codelineno-0-70">70</a></span>
<span class="normal"><a href="#__codelineno-0-71">71</a></span>
<span class="normal"><a href="#__codelineno-0-72">72</a></span>
<span class="normal"><a href="#__codelineno-0-73">73</a></span>
<span class="normal"><a href="#__codelineno-0-74">74</a></span>
<span class="normal"><a href="#__codelineno-0-75">75</a></span>
<span class="normal"><a href="#__codelineno-0-76">76</a></span>
<span class="normal"><a href="#__codelineno-0-77">77</a></span>
<span class="normal"><a href="#__codelineno-0-78">78</a></span>
<span class="normal"><a href="#__codelineno-0-79">79</a></span>
<span class="normal"><a href="#__codelineno-0-80">80</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-20" name="__codelineno-0-20"></a><span class="k">def</span><span class="w"> </span><span class="nf">intersect_clones_in_samples_batch</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-21" name="__codelineno-0-21"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-22" name="__codelineno-0-22"></a><span class="sd">    Calculating frequencies of intersecting clonotypes between multiple repseq samples.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23"></a><span class="sd">    The result of this function may be used for scatterplots of frequencies/counts of </span>
<a id="__codelineno-0-24" name="__codelineno-0-24"></a><span class="sd">    overlapping clonotypes</span>
<a id="__codelineno-0-25" name="__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-27" name="__codelineno-0-27"></a><span class="sd">        clonosets_df (pd.DataFrame): contains three columns - `sample_id` and `filename` columns,</span>
<a id="__codelineno-0-28" name="__codelineno-0-28"></a><span class="sd">            `filename` - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format</span>
<a id="__codelineno-0-29" name="__codelineno-0-29"></a><span class="sd">            sample_id&#39;s should be all unique in this DF</span>
<a id="__codelineno-0-30" name="__codelineno-0-30"></a><span class="sd">        overlap_type (str): possible values are `aa`, `aaV`, `aaVJ`, `nt`, `ntV`, `ntVJ`. aa/nt define which CDR3 sequence</span>
<a id="__codelineno-0-31" name="__codelineno-0-31"></a><span class="sd">            to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments</span>
<a id="__codelineno-0-32" name="__codelineno-0-32"></a><span class="sd">            to decide if clonotypes are equal</span>
<a id="__codelineno-0-33" name="__codelineno-0-33"></a><span class="sd">        by_umi (bool): set `=True` for MiXCR4 clonosets to select count/frequency of clonotypes </span>
<a id="__codelineno-0-34" name="__codelineno-0-34"></a><span class="sd">            in UMI&#39;s if they exist in implemented protocol</span>
<a id="__codelineno-0-35" name="__codelineno-0-35"></a><span class="sd">        by_freq (bool): default is `True` - this means that the intersect metric is frequency of clonotype, </span>
<a id="__codelineno-0-36" name="__codelineno-0-36"></a><span class="sd">            but not its count</span>
<a id="__codelineno-0-37" name="__codelineno-0-37"></a><span class="sd">        only_functional (bool): use only functional clonotypes (do not contain stop codons or</span>
<a id="__codelineno-0-38" name="__codelineno-0-38"></a><span class="sd">            frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to</span>
<a id="__codelineno-0-39" name="__codelineno-0-39"></a><span class="sd">            1 after filtering of non-functional clonotypes</span>
<a id="__codelineno-0-40" name="__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41"></a><span class="sd">    Important: when using particular overlap type, similar clonotypes in one particular clonoset are</span>
<a id="__codelineno-0-42" name="__codelineno-0-42"></a><span class="sd">    combined into one with summation of counts/frequencies.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44"></a><span class="sd">    Returns:</span>
<a id="__codelineno-0-45" name="__codelineno-0-45"></a><span class="sd">        df (pd.DataFrame): dataframe with following columns: `clone`, `sample1_count`, `sample2_count`, `sample1`, `sample2`, `pair`</span>
<a id="__codelineno-0-46" name="__codelineno-0-46"></a><span class="sd">            clone - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric</span>
<a id="__codelineno-0-47" name="__codelineno-0-47"></a><span class="sd">            count columns contain freq/count of the clone in sample</span>
<a id="__codelineno-0-48" name="__codelineno-0-48"></a><span class="sd">            pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</span>
<a id="__codelineno-0-49" name="__codelineno-0-49"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-50" name="__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52"></a>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Intersecting clones in clonosets</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="s2">&quot;-&quot;</span><span class="o">*</span><span class="mi">50</span><span class="p">)</span>
<a id="__codelineno-0-53" name="__codelineno-0-53"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overlap type: </span><span class="si">{</span><span class="n">overlap_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-54" name="__codelineno-0-54"></a>
<a id="__codelineno-0-55" name="__codelineno-0-55"></a>    <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">samples_total</span><span class="p">,</span> <span class="n">two_dataframes</span><span class="p">,</span> <span class="n">sample_list</span><span class="p">,</span> <span class="n">sample_list2</span> <span class="o">=</span> <span class="n">prepare_clonotypes_dfs_for_intersections</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="p">,</span>
<a id="__codelineno-0-56" name="__codelineno-0-56"></a>                                                                                                                        <span class="n">cl_filter</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="p">,</span>
<a id="__codelineno-0-57" name="__codelineno-0-57"></a>                                                                                                                        <span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="n">by_freq</span><span class="p">,</span>
<a id="__codelineno-0-58" name="__codelineno-0-58"></a>                                                                                                                        <span class="n">strict</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59"></a>    <span class="c1"># generating a set of tasks</span>
<a id="__codelineno-0-60" name="__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61"></a>    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-62" name="__codelineno-0-62"></a>
<a id="__codelineno-0-63" name="__codelineno-0-63"></a>    <span class="k">if</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-64" name="__codelineno-0-64"></a>        <span class="k">for</span> <span class="n">sample1</span> <span class="ow">in</span> <span class="n">sample_list</span><span class="p">:</span>
<a id="__codelineno-0-65" name="__codelineno-0-65"></a>            <span class="k">for</span> <span class="n">sample2</span> <span class="ow">in</span> <span class="n">sample_list2</span><span class="p">:</span>
<a id="__codelineno-0-66" name="__codelineno-0-66"></a>                <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">))</span>
<a id="__codelineno-0-67" name="__codelineno-0-67"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-68" name="__codelineno-0-68"></a>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="p">):</span>
<a id="__codelineno-0-69" name="__codelineno-0-69"></a>            <span class="n">sample1</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-70" name="__codelineno-0-70"></a>            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="o">-</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
<a id="__codelineno-0-71" name="__codelineno-0-71"></a>                <span class="n">sample2</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">j</span><span class="o">+</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-72" name="__codelineno-0-72"></a>                <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">))</span>
<a id="__codelineno-0-73" name="__codelineno-0-73"></a>
<a id="__codelineno-0-74" name="__codelineno-0-74"></a>    <span class="n">results</span> <span class="o">=</span> <span class="n">run_parallel_calculation</span><span class="p">(</span><span class="n">intersect_two_clone_dicts</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="s2">&quot;Intersecting clonosets&quot;</span><span class="p">,</span> <span class="n">object_name</span><span class="o">=</span><span class="s2">&quot;pairs&quot;</span><span class="p">)</span>
<a id="__codelineno-0-75" name="__codelineno-0-75"></a>
<a id="__codelineno-0-76" name="__codelineno-0-76"></a>    <span class="c1"># df = pd.concat(results).index.set_names()</span>
<a id="__codelineno-0-77" name="__codelineno-0-77"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">results</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-78" name="__codelineno-0-78"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">split_tuple_clone_column</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">overlap_type</span><span class="p">)</span>
<a id="__codelineno-0-79" name="__codelineno-0-79"></a>
<a id="__codelineno-0-80" name="__codelineno-0-80"></a>    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="intersections.overlap_distances" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">overlap_distances</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s1">&#39;aaV&#39;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;F2&#39;</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Calculating overlap distances between multiple repseq samples using F2 of F metrics
The result of this function may be used for heatmap+clusterization of samples or for MDS plots</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>clonosets_df</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>contains three columns - <code>sample_id</code> and <code>filename</code> columns,
filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format
sample_id's should be all unique in this DF</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>overlap_type</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values are <code>aa</code>, <code>aaV</code>, <code>aaVJ</code>, <code>nt</code>, <code>ntV</code>, <code>ntVJ</code>. aa/nt define which CDR3 sequence
to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments
to decide if clonotypes are equal</p>
              </div>
            </td>
            <td>
                  <code>&#39;aaV&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>mismatches</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The permissible number of single-letter mismatches in clonotypes sequences 
for them to be treated similar, i.e. hamming distance.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>by_umi</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>set =True for MiXCR4 clonosets to select count/frequency of clonotypes 
in UMI's if they exist in implemented protocol</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>metric</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values - <code>F</code>, <code>F2</code> or <code>C</code>. Default <code>F2</code>. <code>F2</code> - sum of sqrt of product of 
similar clonotype frequencies in two clonosets. <code>F</code> - sqrt of the sum of frequency products.
<code>C</code> - total frequency of clonotypes in <code>sample1</code>, that are similar to clonotypes in <code>sample2</code></p>
              </div>
            </td>
            <td>
                  <code>&#39;F2&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>only_functional</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>use only functional clonotypes (do not contain stop codons or
frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to
1 after filtering of non-functional clonotypes</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<details class="important" open>
  <summary>similar clonotypes by <code>overlap_type</code> in one particular clonoset will be combined into one</summary>
  <p>clonotype with sum for count.</p>
</details>

    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>df</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dataframe with following columns: <code>clone</code>, <code>sample1_count</code>, <code>sample2_count</code>, <code>sample1</code>, <code>sample2</code>, <code>pair</code>. 
<code>clone</code> - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric
count columns contain freq/count of the clone in sample
pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-387">387</a></span>
<span class="normal"><a href="#__codelineno-0-388">388</a></span>
<span class="normal"><a href="#__codelineno-0-389">389</a></span>
<span class="normal"><a href="#__codelineno-0-390">390</a></span>
<span class="normal"><a href="#__codelineno-0-391">391</a></span>
<span class="normal"><a href="#__codelineno-0-392">392</a></span>
<span class="normal"><a href="#__codelineno-0-393">393</a></span>
<span class="normal"><a href="#__codelineno-0-394">394</a></span>
<span class="normal"><a href="#__codelineno-0-395">395</a></span>
<span class="normal"><a href="#__codelineno-0-396">396</a></span>
<span class="normal"><a href="#__codelineno-0-397">397</a></span>
<span class="normal"><a href="#__codelineno-0-398">398</a></span>
<span class="normal"><a href="#__codelineno-0-399">399</a></span>
<span class="normal"><a href="#__codelineno-0-400">400</a></span>
<span class="normal"><a href="#__codelineno-0-401">401</a></span>
<span class="normal"><a href="#__codelineno-0-402">402</a></span>
<span class="normal"><a href="#__codelineno-0-403">403</a></span>
<span class="normal"><a href="#__codelineno-0-404">404</a></span>
<span class="normal"><a href="#__codelineno-0-405">405</a></span>
<span class="normal"><a href="#__codelineno-0-406">406</a></span>
<span class="normal"><a href="#__codelineno-0-407">407</a></span>
<span class="normal"><a href="#__codelineno-0-408">408</a></span>
<span class="normal"><a href="#__codelineno-0-409">409</a></span>
<span class="normal"><a href="#__codelineno-0-410">410</a></span>
<span class="normal"><a href="#__codelineno-0-411">411</a></span>
<span class="normal"><a href="#__codelineno-0-412">412</a></span>
<span class="normal"><a href="#__codelineno-0-413">413</a></span>
<span class="normal"><a href="#__codelineno-0-414">414</a></span>
<span class="normal"><a href="#__codelineno-0-415">415</a></span>
<span class="normal"><a href="#__codelineno-0-416">416</a></span>
<span class="normal"><a href="#__codelineno-0-417">417</a></span>
<span class="normal"><a href="#__codelineno-0-418">418</a></span>
<span class="normal"><a href="#__codelineno-0-419">419</a></span>
<span class="normal"><a href="#__codelineno-0-420">420</a></span>
<span class="normal"><a href="#__codelineno-0-421">421</a></span>
<span class="normal"><a href="#__codelineno-0-422">422</a></span>
<span class="normal"><a href="#__codelineno-0-423">423</a></span>
<span class="normal"><a href="#__codelineno-0-424">424</a></span>
<span class="normal"><a href="#__codelineno-0-425">425</a></span>
<span class="normal"><a href="#__codelineno-0-426">426</a></span>
<span class="normal"><a href="#__codelineno-0-427">427</a></span>
<span class="normal"><a href="#__codelineno-0-428">428</a></span>
<span class="normal"><a href="#__codelineno-0-429">429</a></span>
<span class="normal"><a href="#__codelineno-0-430">430</a></span>
<span class="normal"><a href="#__codelineno-0-431">431</a></span>
<span class="normal"><a href="#__codelineno-0-432">432</a></span>
<span class="normal"><a href="#__codelineno-0-433">433</a></span>
<span class="normal"><a href="#__codelineno-0-434">434</a></span>
<span class="normal"><a href="#__codelineno-0-435">435</a></span>
<span class="normal"><a href="#__codelineno-0-436">436</a></span>
<span class="normal"><a href="#__codelineno-0-437">437</a></span>
<span class="normal"><a href="#__codelineno-0-438">438</a></span>
<span class="normal"><a href="#__codelineno-0-439">439</a></span>
<span class="normal"><a href="#__codelineno-0-440">440</a></span>
<span class="normal"><a href="#__codelineno-0-441">441</a></span>
<span class="normal"><a href="#__codelineno-0-442">442</a></span>
<span class="normal"><a href="#__codelineno-0-443">443</a></span>
<span class="normal"><a href="#__codelineno-0-444">444</a></span>
<span class="normal"><a href="#__codelineno-0-445">445</a></span>
<span class="normal"><a href="#__codelineno-0-446">446</a></span>
<span class="normal"><a href="#__codelineno-0-447">447</a></span>
<span class="normal"><a href="#__codelineno-0-448">448</a></span>
<span class="normal"><a href="#__codelineno-0-449">449</a></span>
<span class="normal"><a href="#__codelineno-0-450">450</a></span>
<span class="normal"><a href="#__codelineno-0-451">451</a></span>
<span class="normal"><a href="#__codelineno-0-452">452</a></span>
<span class="normal"><a href="#__codelineno-0-453">453</a></span>
<span class="normal"><a href="#__codelineno-0-454">454</a></span>
<span class="normal"><a href="#__codelineno-0-455">455</a></span>
<span class="normal"><a href="#__codelineno-0-456">456</a></span>
<span class="normal"><a href="#__codelineno-0-457">457</a></span>
<span class="normal"><a href="#__codelineno-0-458">458</a></span>
<span class="normal"><a href="#__codelineno-0-459">459</a></span>
<span class="normal"><a href="#__codelineno-0-460">460</a></span>
<span class="normal"><a href="#__codelineno-0-461">461</a></span>
<span class="normal"><a href="#__codelineno-0-462">462</a></span>
<span class="normal"><a href="#__codelineno-0-463">463</a></span>
<span class="normal"><a href="#__codelineno-0-464">464</a></span>
<span class="normal"><a href="#__codelineno-0-465">465</a></span>
<span class="normal"><a href="#__codelineno-0-466">466</a></span>
<span class="normal"><a href="#__codelineno-0-467">467</a></span>
<span class="normal"><a href="#__codelineno-0-468">468</a></span>
<span class="normal"><a href="#__codelineno-0-469">469</a></span>
<span class="normal"><a href="#__codelineno-0-470">470</a></span>
<span class="normal"><a href="#__codelineno-0-471">471</a></span>
<span class="normal"><a href="#__codelineno-0-472">472</a></span>
<span class="normal"><a href="#__codelineno-0-473">473</a></span>
<span class="normal"><a href="#__codelineno-0-474">474</a></span>
<span class="normal"><a href="#__codelineno-0-475">475</a></span>
<span class="normal"><a href="#__codelineno-0-476">476</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-387" name="__codelineno-0-387"></a><span class="k">def</span><span class="w"> </span><span class="nf">overlap_distances</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;F2&quot;</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-388" name="__codelineno-0-388"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-389" name="__codelineno-0-389"></a><span class="sd">    Calculating overlap distances between multiple repseq samples using F2 of F metrics</span>
<a id="__codelineno-0-390" name="__codelineno-0-390"></a><span class="sd">    The result of this function may be used for heatmap+clusterization of samples or for MDS plots</span>
<a id="__codelineno-0-391" name="__codelineno-0-391"></a>
<a id="__codelineno-0-392" name="__codelineno-0-392"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-393" name="__codelineno-0-393"></a><span class="sd">        clonosets_df (pd.DataFrame): contains three columns - `sample_id` and `filename` columns,</span>
<a id="__codelineno-0-394" name="__codelineno-0-394"></a><span class="sd">            filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format</span>
<a id="__codelineno-0-395" name="__codelineno-0-395"></a><span class="sd">            sample_id&#39;s should be all unique in this DF</span>
<a id="__codelineno-0-396" name="__codelineno-0-396"></a><span class="sd">        overlap_type (str): possible values are `aa`, `aaV`, `aaVJ`, `nt`, `ntV`, `ntVJ`. aa/nt define which CDR3 sequence</span>
<a id="__codelineno-0-397" name="__codelineno-0-397"></a><span class="sd">            to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments</span>
<a id="__codelineno-0-398" name="__codelineno-0-398"></a><span class="sd">            to decide if clonotypes are equal</span>
<a id="__codelineno-0-399" name="__codelineno-0-399"></a><span class="sd">        mismatches (int): The permissible number of single-letter mismatches in clonotypes sequences </span>
<a id="__codelineno-0-400" name="__codelineno-0-400"></a><span class="sd">            for them to be treated similar, i.e. hamming distance.</span>
<a id="__codelineno-0-401" name="__codelineno-0-401"></a><span class="sd">        by_umi (bool): set =True for MiXCR4 clonosets to select count/frequency of clonotypes </span>
<a id="__codelineno-0-402" name="__codelineno-0-402"></a><span class="sd">            in UMI&#39;s if they exist in implemented protocol</span>
<a id="__codelineno-0-403" name="__codelineno-0-403"></a><span class="sd">        metric (str): possible values - `F`, `F2` or `C`. Default `F2`. `F2` - sum of sqrt of product of </span>
<a id="__codelineno-0-404" name="__codelineno-0-404"></a><span class="sd">            similar clonotype frequencies in two clonosets. `F` - sqrt of the sum of frequency products.</span>
<a id="__codelineno-0-405" name="__codelineno-0-405"></a><span class="sd">            `C` - total frequency of clonotypes in `sample1`, that are similar to clonotypes in `sample2`</span>
<a id="__codelineno-0-406" name="__codelineno-0-406"></a><span class="sd">        only_functional (bool): use only functional clonotypes (do not contain stop codons or</span>
<a id="__codelineno-0-407" name="__codelineno-0-407"></a><span class="sd">            frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to</span>
<a id="__codelineno-0-408" name="__codelineno-0-408"></a><span class="sd">            1 after filtering of non-functional clonotypes</span>
<a id="__codelineno-0-409" name="__codelineno-0-409"></a>
<a id="__codelineno-0-410" name="__codelineno-0-410"></a><span class="sd">    Important: similar clonotypes by `overlap_type` in one particular clonoset will be combined into one</span>
<a id="__codelineno-0-411" name="__codelineno-0-411"></a><span class="sd">        clonotype with sum for count.</span>
<a id="__codelineno-0-412" name="__codelineno-0-412"></a>
<a id="__codelineno-0-413" name="__codelineno-0-413"></a><span class="sd">    Returns:</span>
<a id="__codelineno-0-414" name="__codelineno-0-414"></a><span class="sd">        df (pd.DataFrame): dataframe with following columns: `clone`, `sample1_count`, `sample2_count`, `sample1`, `sample2`, `pair`. </span>
<a id="__codelineno-0-415" name="__codelineno-0-415"></a><span class="sd">            `clone` - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric</span>
<a id="__codelineno-0-416" name="__codelineno-0-416"></a><span class="sd">            count columns contain freq/count of the clone in sample</span>
<a id="__codelineno-0-417" name="__codelineno-0-417"></a><span class="sd">            pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</span>
<a id="__codelineno-0-418" name="__codelineno-0-418"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-419" name="__codelineno-0-419"></a>
<a id="__codelineno-0-420" name="__codelineno-0-420"></a>
<a id="__codelineno-0-421" name="__codelineno-0-421"></a>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Intersecting clones in clonosets</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="s2">&quot;-&quot;</span><span class="o">*</span><span class="mi">50</span><span class="p">)</span>
<a id="__codelineno-0-422" name="__codelineno-0-422"></a>    <span class="n">aa</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span> <span class="o">=</span> <span class="n">overlap_type_to_flags</span><span class="p">(</span><span class="n">overlap_type</span><span class="p">)</span>
<a id="__codelineno-0-423" name="__codelineno-0-423"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overlap type: </span><span class="si">{</span><span class="n">overlap_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-424" name="__codelineno-0-424"></a>
<a id="__codelineno-0-425" name="__codelineno-0-425"></a>    <span class="n">metric</span> <span class="o">=</span> <span class="n">metric</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<a id="__codelineno-0-426" name="__codelineno-0-426"></a>    <span class="n">metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;F&quot;</span><span class="p">,</span> <span class="s2">&quot;F2&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;BC&quot;</span><span class="p">,</span> <span class="s2">&quot;J&quot;</span><span class="p">,</span> <span class="s2">&quot;JSD&quot;</span><span class="p">]</span>
<a id="__codelineno-0-427" name="__codelineno-0-427"></a>    <span class="n">mismatch_metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;F&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">]</span>
<a id="__codelineno-0-428" name="__codelineno-0-428"></a>    <span class="n">non_symmetry_metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">]</span>
<a id="__codelineno-0-429" name="__codelineno-0-429"></a>    <span class="n">frequency_metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;F&quot;</span><span class="p">,</span> <span class="s2">&quot;F2&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">]</span>
<a id="__codelineno-0-430" name="__codelineno-0-430"></a>
<a id="__codelineno-0-431" name="__codelineno-0-431"></a>
<a id="__codelineno-0-432" name="__codelineno-0-432"></a>    <span class="k">if</span> <span class="n">metric</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">metrics</span><span class="p">:</span>
<a id="__codelineno-0-433" name="__codelineno-0-433"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Metric </span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2"> is not supported. Possible values: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-434" name="__codelineno-0-434"></a>
<a id="__codelineno-0-435" name="__codelineno-0-435"></a>    <span class="k">if</span> <span class="n">mismatches</span> <span class="ow">and</span> <span class="n">metric</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mismatch_metrics</span><span class="p">:</span>
<a id="__codelineno-0-436" name="__codelineno-0-436"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Metric </span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2"> does not allow mismatches. Mismatches only possible for: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mismatch_metrics</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-437" name="__codelineno-0-437"></a>
<a id="__codelineno-0-438" name="__codelineno-0-438"></a>    <span class="n">by_freq</span> <span class="o">=</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">frequency_metrics</span>
<a id="__codelineno-0-439" name="__codelineno-0-439"></a>
<a id="__codelineno-0-440" name="__codelineno-0-440"></a>    <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">samples_total</span><span class="p">,</span> <span class="n">two_dataframes</span><span class="p">,</span> <span class="n">sample_list</span><span class="p">,</span> <span class="n">sample_list2</span> <span class="o">=</span> <span class="n">prepare_clonotypes_dfs_for_intersections</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="p">,</span>
<a id="__codelineno-0-441" name="__codelineno-0-441"></a>                                                                                                                        <span class="n">cl_filter</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="p">,</span>
<a id="__codelineno-0-442" name="__codelineno-0-442"></a>                                                                                                                        <span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="n">by_freq</span><span class="p">)</span>
<a id="__codelineno-0-443" name="__codelineno-0-443"></a>
<a id="__codelineno-0-444" name="__codelineno-0-444"></a>    <span class="c1"># generating a set of tasks</span>
<a id="__codelineno-0-445" name="__codelineno-0-445"></a>
<a id="__codelineno-0-446" name="__codelineno-0-446"></a>    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-447" name="__codelineno-0-447"></a>
<a id="__codelineno-0-448" name="__codelineno-0-448"></a>    <span class="k">if</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-449" name="__codelineno-0-449"></a>        <span class="k">for</span> <span class="n">sample1</span> <span class="ow">in</span> <span class="n">sample_list</span><span class="p">:</span>
<a id="__codelineno-0-450" name="__codelineno-0-450"></a>            <span class="k">for</span> <span class="n">sample2</span> <span class="ow">in</span> <span class="n">sample_list2</span><span class="p">:</span>
<a id="__codelineno-0-451" name="__codelineno-0-451"></a>                <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">,</span> <span class="n">metric</span><span class="p">))</span>        
<a id="__codelineno-0-452" name="__codelineno-0-452"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-453" name="__codelineno-0-453"></a>        <span class="k">if</span> <span class="n">metric</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">non_symmetry_metrics</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-454" name="__codelineno-0-454"></a>            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="p">):</span>
<a id="__codelineno-0-455" name="__codelineno-0-455"></a>                <span class="n">sample1</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-456" name="__codelineno-0-456"></a>                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="o">-</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
<a id="__codelineno-0-457" name="__codelineno-0-457"></a>                    <span class="n">sample2</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">j</span><span class="o">+</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-458" name="__codelineno-0-458"></a>                    <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">,</span> <span class="n">metric</span><span class="p">))</span>
<a id="__codelineno-0-459" name="__codelineno-0-459"></a>                <span class="k">if</span> <span class="n">metric</span> <span class="o">==</span> <span class="s2">&quot;F2&quot;</span><span class="p">:</span>
<a id="__codelineno-0-460" name="__codelineno-0-460"></a>                    <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample1</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">,</span> <span class="n">metric</span><span class="p">))</span>
<a id="__codelineno-0-461" name="__codelineno-0-461"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-462" name="__codelineno-0-462"></a>            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="p">):</span>
<a id="__codelineno-0-463" name="__codelineno-0-463"></a>                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="p">):</span>
<a id="__codelineno-0-464" name="__codelineno-0-464"></a>                    <span class="n">sample1</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-465" name="__codelineno-0-465"></a>                    <span class="n">sample2</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
<a id="__codelineno-0-466" name="__codelineno-0-466"></a>                    <span class="k">if</span> <span class="n">sample1</span> <span class="o">!=</span> <span class="n">sample2</span><span class="p">:</span>
<a id="__codelineno-0-467" name="__codelineno-0-467"></a>                        <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">,</span> <span class="n">metric</span><span class="p">))</span>
<a id="__codelineno-0-468" name="__codelineno-0-468"></a>
<a id="__codelineno-0-469" name="__codelineno-0-469"></a>
<a id="__codelineno-0-470" name="__codelineno-0-470"></a>    <span class="c1"># run calculation in parallel</span>
<a id="__codelineno-0-471" name="__codelineno-0-471"></a>    <span class="n">result_list</span> <span class="o">=</span> <span class="n">run_parallel_calculation</span><span class="p">(</span><span class="n">overlap_metric_two_clone_dicts</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="s2">&quot;Intersecting clonosets&quot;</span><span class="p">,</span> <span class="n">object_name</span><span class="o">=</span><span class="s2">&quot;pairs&quot;</span><span class="p">)</span>
<a id="__codelineno-0-472" name="__codelineno-0-472"></a>
<a id="__codelineno-0-473" name="__codelineno-0-473"></a>    <span class="k">if</span> <span class="ow">not</span> <span class="n">two_dataframes</span> <span class="ow">and</span> <span class="n">metric</span> <span class="o">!=</span> <span class="s2">&quot;C&quot;</span><span class="p">:</span>
<a id="__codelineno-0-474" name="__codelineno-0-474"></a>        <span class="n">result_list</span> <span class="o">=</span> <span class="n">result_list</span> <span class="o">+</span> <span class="p">[(</span><span class="n">result</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">result</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">result_list</span><span class="p">]</span>
<a id="__codelineno-0-475" name="__codelineno-0-475"></a>    <span class="n">overlap_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">result_list</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sample1&quot;</span><span class="p">,</span> <span class="s2">&quot;sample2&quot;</span><span class="p">,</span> <span class="n">metric</span><span class="o">.</span><span class="n">lower</span><span class="p">()])</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="s2">&quot;sample1&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sample2&quot;</span><span class="p">],</span> <span class="n">values</span><span class="o">=</span><span class="n">metric</span><span class="o">.</span><span class="n">lower</span><span class="p">())</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;sample1&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-476" name="__codelineno-0-476"></a>    <span class="k">return</span> <span class="n">overlap_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="intersections.prepare_clonotypes_dfs_for_intersections" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">prepare_clonotypes_dfs_for_intersections</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="p">,</span> <span class="n">cl_filter</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="p">,</span> <span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">



<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>clonosets_df</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>clonosets_df2</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>cl_filter</code></td>
            <td>
                  <code><span title="repseq.clone_filter.Filter">Filter</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>cl_filter2</code></td>
            <td>
                  <code><span title="repseq.clone_filter.Filter">Filter</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>overlap_type</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>by_freq</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em>. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>clonoset_lists</code></td>            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dict of </p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>samples_total</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>two_dataframes</code></td>            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>sample_list</code></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>sample_list2</code></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-547">547</a></span>
<span class="normal"><a href="#__codelineno-0-548">548</a></span>
<span class="normal"><a href="#__codelineno-0-549">549</a></span>
<span class="normal"><a href="#__codelineno-0-550">550</a></span>
<span class="normal"><a href="#__codelineno-0-551">551</a></span>
<span class="normal"><a href="#__codelineno-0-552">552</a></span>
<span class="normal"><a href="#__codelineno-0-553">553</a></span>
<span class="normal"><a href="#__codelineno-0-554">554</a></span>
<span class="normal"><a href="#__codelineno-0-555">555</a></span>
<span class="normal"><a href="#__codelineno-0-556">556</a></span>
<span class="normal"><a href="#__codelineno-0-557">557</a></span>
<span class="normal"><a href="#__codelineno-0-558">558</a></span>
<span class="normal"><a href="#__codelineno-0-559">559</a></span>
<span class="normal"><a href="#__codelineno-0-560">560</a></span>
<span class="normal"><a href="#__codelineno-0-561">561</a></span>
<span class="normal"><a href="#__codelineno-0-562">562</a></span>
<span class="normal"><a href="#__codelineno-0-563">563</a></span>
<span class="normal"><a href="#__codelineno-0-564">564</a></span>
<span class="normal"><a href="#__codelineno-0-565">565</a></span>
<span class="normal"><a href="#__codelineno-0-566">566</a></span>
<span class="normal"><a href="#__codelineno-0-567">567</a></span>
<span class="normal"><a href="#__codelineno-0-568">568</a></span>
<span class="normal"><a href="#__codelineno-0-569">569</a></span>
<span class="normal"><a href="#__codelineno-0-570">570</a></span>
<span class="normal"><a href="#__codelineno-0-571">571</a></span>
<span class="normal"><a href="#__codelineno-0-572">572</a></span>
<span class="normal"><a href="#__codelineno-0-573">573</a></span>
<span class="normal"><a href="#__codelineno-0-574">574</a></span>
<span class="normal"><a href="#__codelineno-0-575">575</a></span>
<span class="normal"><a href="#__codelineno-0-576">576</a></span>
<span class="normal"><a href="#__codelineno-0-577">577</a></span>
<span class="normal"><a href="#__codelineno-0-578">578</a></span>
<span class="normal"><a href="#__codelineno-0-579">579</a></span>
<span class="normal"><a href="#__codelineno-0-580">580</a></span>
<span class="normal"><a href="#__codelineno-0-581">581</a></span>
<span class="normal"><a href="#__codelineno-0-582">582</a></span>
<span class="normal"><a href="#__codelineno-0-583">583</a></span>
<span class="normal"><a href="#__codelineno-0-584">584</a></span>
<span class="normal"><a href="#__codelineno-0-585">585</a></span>
<span class="normal"><a href="#__codelineno-0-586">586</a></span>
<span class="normal"><a href="#__codelineno-0-587">587</a></span>
<span class="normal"><a href="#__codelineno-0-588">588</a></span>
<span class="normal"><a href="#__codelineno-0-589">589</a></span>
<span class="normal"><a href="#__codelineno-0-590">590</a></span>
<span class="normal"><a href="#__codelineno-0-591">591</a></span>
<span class="normal"><a href="#__codelineno-0-592">592</a></span>
<span class="normal"><a href="#__codelineno-0-593">593</a></span>
<span class="normal"><a href="#__codelineno-0-594">594</a></span>
<span class="normal"><a href="#__codelineno-0-595">595</a></span>
<span class="normal"><a href="#__codelineno-0-596">596</a></span>
<span class="normal"><a href="#__codelineno-0-597">597</a></span>
<span class="normal"><a href="#__codelineno-0-598">598</a></span>
<span class="normal"><a href="#__codelineno-0-599">599</a></span>
<span class="normal"><a href="#__codelineno-0-600">600</a></span>
<span class="normal"><a href="#__codelineno-0-601">601</a></span>
<span class="normal"><a href="#__codelineno-0-602">602</a></span>
<span class="normal"><a href="#__codelineno-0-603">603</a></span>
<span class="normal"><a href="#__codelineno-0-604">604</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-547" name="__codelineno-0-547"></a><span class="k">def</span><span class="w"> </span><span class="nf">prepare_clonotypes_dfs_for_intersections</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="p">,</span> <span class="n">cl_filter</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="p">,</span> <span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-548" name="__codelineno-0-548"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-549" name="__codelineno-0-549"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-550" name="__codelineno-0-550"></a><span class="sd">        clonosets_df (pd.DataFrame): _description_</span>
<a id="__codelineno-0-551" name="__codelineno-0-551"></a><span class="sd">        clonosets_df2 (pd.DataFrame): _description_</span>
<a id="__codelineno-0-552" name="__codelineno-0-552"></a><span class="sd">        cl_filter (Filter): _description_</span>
<a id="__codelineno-0-553" name="__codelineno-0-553"></a><span class="sd">        cl_filter2 (Filter): _description_</span>
<a id="__codelineno-0-554" name="__codelineno-0-554"></a><span class="sd">        overlap_type (str): _description_</span>
<a id="__codelineno-0-555" name="__codelineno-0-555"></a><span class="sd">        by_freq (bool, optional): _description_. Defaults to True.</span>
<a id="__codelineno-0-556" name="__codelineno-0-556"></a>
<a id="__codelineno-0-557" name="__codelineno-0-557"></a><span class="sd">    Raises:</span>
<a id="__codelineno-0-558" name="__codelineno-0-558"></a><span class="sd">        ValueError: _description_</span>
<a id="__codelineno-0-559" name="__codelineno-0-559"></a><span class="sd">        ValueError: _description_</span>
<a id="__codelineno-0-560" name="__codelineno-0-560"></a>
<a id="__codelineno-0-561" name="__codelineno-0-561"></a><span class="sd">    Returns:</span>
<a id="__codelineno-0-562" name="__codelineno-0-562"></a><span class="sd">        clonoset_lists (dict): dict of </span>
<a id="__codelineno-0-563" name="__codelineno-0-563"></a><span class="sd">        samples_total (int): </span>
<a id="__codelineno-0-564" name="__codelineno-0-564"></a><span class="sd">        two_dataframes (bool):</span>
<a id="__codelineno-0-565" name="__codelineno-0-565"></a><span class="sd">        sample_list (list):</span>
<a id="__codelineno-0-566" name="__codelineno-0-566"></a><span class="sd">        sample_list2 (list):</span>
<a id="__codelineno-0-567" name="__codelineno-0-567"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-568" name="__codelineno-0-568"></a>    <span class="c1"># output:</span>
<a id="__codelineno-0-569" name="__codelineno-0-569"></a>    <span class="c1">### clonoset_lists</span>
<a id="__codelineno-0-570" name="__codelineno-0-570"></a>
<a id="__codelineno-0-571" name="__codelineno-0-571"></a>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">clonosets_df</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">):</span>
<a id="__codelineno-0-572" name="__codelineno-0-572"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Input clonosets in DataFrame have non-unique sample_id&#39;s&quot;</span><span class="p">)</span>
<a id="__codelineno-0-573" name="__codelineno-0-573"></a>    <span class="n">clonosets_df_1</span> <span class="o">=</span> <span class="n">clonosets_df</span><span class="p">[[</span><span class="s2">&quot;sample_id&quot;</span><span class="p">,</span> <span class="s2">&quot;filename&quot;</span><span class="p">]]</span>
<a id="__codelineno-0-574" name="__codelineno-0-574"></a>    <span class="n">two_dataframes</span> <span class="o">=</span> <span class="kc">False</span>
<a id="__codelineno-0-575" name="__codelineno-0-575"></a>    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">clonosets_df2</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<a id="__codelineno-0-576" name="__codelineno-0-576"></a>        <span class="n">two_dataframes</span> <span class="o">=</span> <span class="kc">True</span>
<a id="__codelineno-0-577" name="__codelineno-0-577"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">clonosets_df2</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">clonosets_df2</span><span class="p">):</span>
<a id="__codelineno-0-578" name="__codelineno-0-578"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Input clonosets in DataFrame2 have non-unique sample_id&#39;s&quot;</span><span class="p">)</span>
<a id="__codelineno-0-579" name="__codelineno-0-579"></a>        <span class="n">clonosets_df_2</span> <span class="o">=</span> <span class="n">clonosets_df2</span><span class="p">[[</span><span class="s2">&quot;sample_id&quot;</span><span class="p">,</span> <span class="s2">&quot;filename&quot;</span><span class="p">]]</span>
<a id="__codelineno-0-580" name="__codelineno-0-580"></a>        <span class="n">intersecting_sample_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">clonosets_df2</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">clonosets_df</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
<a id="__codelineno-0-581" name="__codelineno-0-581"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">intersecting_sample_ids</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">cl_filter2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-582" name="__codelineno-0-582"></a>            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;WARNING! Some samples have the same sample_id in two sample_df&#39;s. The second filter will be applied to common samples&quot;</span><span class="p">)</span>
<a id="__codelineno-0-583" name="__codelineno-0-583"></a>
<a id="__codelineno-0-584" name="__codelineno-0-584"></a>
<a id="__codelineno-0-585" name="__codelineno-0-585"></a>    <span class="c1"># converting clonosets to compact lists of clonotypes separated by CDR3 lengths to dictionary based on overlap type and count/freq/umi</span>
<a id="__codelineno-0-586" name="__codelineno-0-586"></a>    <span class="n">clonoset_lists</span> <span class="o">=</span> <span class="n">convert_clonosets_to_compact_dicts</span><span class="p">(</span><span class="n">clonosets_df_1</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">cl_filter</span><span class="p">,</span>
<a id="__codelineno-0-587" name="__codelineno-0-587"></a>                                                        <span class="n">overlap_type</span><span class="o">=</span><span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="n">by_freq</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="n">strict</span><span class="p">)</span>
<a id="__codelineno-0-588" name="__codelineno-0-588"></a>    <span class="k">if</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-589" name="__codelineno-0-589"></a>        <span class="k">if</span> <span class="n">cl_filter2</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-590" name="__codelineno-0-590"></a>            <span class="n">cl_filter2</span> <span class="o">=</span> <span class="n">cl_filter</span>
<a id="__codelineno-0-591" name="__codelineno-0-591"></a>        <span class="n">clonoset_lists_2</span> <span class="o">=</span> <span class="n">convert_clonosets_to_compact_dicts</span><span class="p">(</span><span class="n">clonosets_df_2</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">cl_filter2</span><span class="p">,</span>
<a id="__codelineno-0-592" name="__codelineno-0-592"></a>                                                        <span class="n">overlap_type</span><span class="o">=</span><span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="n">by_freq</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="n">strict</span><span class="p">)</span>
<a id="__codelineno-0-593" name="__codelineno-0-593"></a>        <span class="n">clonoset_lists</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">clonoset_lists_2</span><span class="p">)</span>
<a id="__codelineno-0-594" name="__codelineno-0-594"></a>
<a id="__codelineno-0-595" name="__codelineno-0-595"></a>    <span class="n">samples_total</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">clonosets_df_1</span><span class="p">)</span>
<a id="__codelineno-0-596" name="__codelineno-0-596"></a>    <span class="k">if</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-597" name="__codelineno-0-597"></a>        <span class="n">samples_total</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">clonosets_df_1</span><span class="p">,</span> <span class="n">clonosets_df_2</span><span class="p">]))</span>
<a id="__codelineno-0-598" name="__codelineno-0-598"></a>
<a id="__codelineno-0-599" name="__codelineno-0-599"></a>    <span class="n">sample_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">clonosets_df_1</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;sample_id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sample_id</span><span class="p">)</span>
<a id="__codelineno-0-600" name="__codelineno-0-600"></a>    <span class="n">sample_list2</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-601" name="__codelineno-0-601"></a>    <span class="k">if</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-602" name="__codelineno-0-602"></a>        <span class="n">sample_list2</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">clonosets_df_2</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;sample_id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sample_id</span><span class="p">)</span>
<a id="__codelineno-0-603" name="__codelineno-0-603"></a>
<a id="__codelineno-0-604" name="__codelineno-0-604"></a>    <span class="k">return</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">samples_total</span><span class="p">,</span> <span class="n">two_dataframes</span><span class="p">,</span> <span class="n">sample_list</span><span class="p">,</span> <span class="n">sample_list2</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy; 2023 <a href="https://github.com/mmjmike" target="_blank" rel="noopener">Mikhail Myshkin</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/mmjmike" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "navigation.sections", "navigation.path", "toc.integrate", "navigation.top", "search.suggest", "search.highlight", "content.tabs.link", "content.code.annotation", "content.code.copy"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.f1b6f286.min.js"></script>
      
    
  </body>
</html>